src/jltypes.c

```diff
@@ -1575,6 +1575,12 @@ static unsigned type_hash(jl_value_t *kj, int *failed) JL_NOTSAFEPOINT
     }
 }
 
+JL_DLLEXPORT uintptr_t jl_type_hash(jl_value_t *v) JL_NOTSAFEPOINT
+{
+    int failed = 0;
+    return type_hash(v, &failed);
```

Comment:
It's not clear to me why it's ok to ignore `failed` here. Maybe a comment about it? If it shouldn't ever happen, then perhaps an assertion that hasn't been set?

Reply:
Yeah, fair point. When `failed` is set, it means the type hash was determined unstable and the returned hash value is 0. That's probably not great, perhaps we should fall back to using `objectid`.

Does something like this seem more reasonable?
```diff
diff --git a/src/jltypes.c b/src/jltypes.c
index f6c8236376..04d0e7cc96 100644
--- a/src/jltypes.c
+++ b/src/jltypes.c
@@ -1578,7 +1578,13 @@ static unsigned type_hash(jl_value_t *kj, int *failed) JL_NOTSAFEPOINT
 JL_DLLEXPORT uintptr_t jl_type_hash(jl_value_t *v) JL_NOTSAFEPOINT
 {
     int failed = 0;
-    return type_hash(v, &failed);
+    unsigned hash = type_hash(v, &failed);
+    // Fall back to using the object ID if the type hash is unstable
+    if (failed) {
+        jl_value_t *uw = jl_is_unionall(v) ? jl_unwrap_unionall(v) : v;
+        hash = jl_object_id(uw);
+    }
+    return hash;
 }

 static unsigned typekey_hash(jl_typename_t *tn, jl_value_t **key, size_t n, int nofail) JL_NOTSAFEPOINT
```

Reply:
No, you are supposed to ignore failed here. It is needed for a different part of the algorithm, but returning 0 is the correct answer